### ISSUE TITLE
[Beta] Handle errors thrown anywhere in chained Future methods

### ISSUE THREAD

**Issue by bensyverson** [vapor/async#68]
## Summary
With Async, it's idiomatic to chain methods. However, in methods that create new Futures (such as Fluent's `Model.save()`), any previous errors thrown seem to be discarded. This means calls to `Abort()` will not always abort the request.

## Details
It's highly desirable to be able to do something like the following:

```swift
return try User.query(on: req)
		.filter(\User.username == userReq.id)
		.count()
		.makeSureCountIsZero()
		.saveUser(userRequest: userReq, on: req)
```

Where `makeSureCountIsZero` throws if `count != 0`, and `saveUser` is like this (simplified for clarity):

```swift
extension Future where T == Bool {
  func saveUser(userRequest: UserRequest, on connectable: Request) throws -> Future<User> {
    let newUser = User(username: name)
    return newUser.save(on: connectable)
  }
}
```

However, `Model.save()` is creating a brand new Future, which I believe is discarding any previous exceptions/errors on the previous Future.

## Simplified Example
In `TodoController.swift`:

```swift
extension Future where T == String {
  func returnString() throws -> Future<String> {
    return Future("✨ everything is perfect ✨")
  }
}

final class TodoController {
  func getThrow(_ req: Request) throws -> Future<String> {
    return try Todo.query(on: req)
      .filter(\Todo.title == "title")
      .count()
      .map(to: String.self) { _ in
        throw Abort(.badRequest, reason: "Everything is not perfect")
      }
      .returnString()
  }
}
```

Calling the `getThrow` method should return a 400, but instead we get a 200 OK and successful response from `returnString`

## Steps to Reproduce
1. Download [FutureAsync.zip](https://github.com/vapor/async/files/1713739/FutureAsync.zip)
2. Unpack and build (may require `2018-01-30` toolchain)
3. Run and request `GET /throw`

## Workarounds / Regression
I think the correct thing to do is actually to `flatMap` the future before returning a new future, as in the following, which does `throw` the Abort correctly.

```swift
extension Future where T == String {
  func returnString() throws -> Future<String> {
    return self.flatMap(to: String.self) { _ in
      return Future("✨ everything is perfect ✨")
    }
  }
}
```

However, many users will not realize this, and their errors will be silently caught and ignored. Obviously that could be extremely dangerous. I'm hoping there's a way to force the error to be bubbled up. If there isn't, hopefully there's a way to catch this user error at compile time.

**Comment by twof** [comment]
This is most likely the issue mentioned in the comments of 
https://github.com/vapor/async/pull/63

**Comment by bensyverson** [comment]
@twof Yeah, I think so. It's cool for #63 to be spun out to a separate helper package, but to me, this issue (#68) is a bug with Async itself.
